star-wars-app/Network/SWApi.swift

```diff
@@ -2,32 +2,39 @@
 //  SWApi.swift
 //  star-wars-app
 //
-//  Created by Michał Nowak on 02/01/2020.
+//  Created by Michał Nowak on 10/01/2020.
 //  Copyright © 2020 Michal Nowak. All rights reserved.
 //
 
 import Foundation
-import Moya
-import RxMoya
 import RxSwift
 
-class SWApi {
-    let provider = MoyaProvider<SW>()
-    
-    public func request(_ endpoint: SW) -> Observable<Response> {
-           return self.provider.rx.request(endpoint).asObservable().flatMap({ (response) -> Observable<Response> in
-               return response.catchAPIErrorIfNeeded()
-           })
-       }
+struct APIResponse {
+    let data: Data?
+    let response: URLResponse?
 }
 
-extension Response {
-    func catchAPIErrorIfNeeded() -> Observable<Response> {
-        do {
-            let response = try self.filterSuccessfulStatusCodes()
-            return Observable.just(response)
-        } catch {
-            return Observable.just(self)
+class SWApi {
+    private let baseURL = URL(string: "https://swapi.co/api/")!
+    
+    func request(apiRequest: APIRequest) -> Observable<APIResponse> {
+        return Observable<APIResponse>.create { observer in
+            let request = apiRequest.request(with: self.baseURL)
+            let task = URLSession.shared.dataTask(with: request) { (data, response, error) in
+                if error != nil {
```

Comment:
that's a detail. Do you know conditional unwrapping in swift?
you can do it like this:
```swift
if let error = error {
    observer.onError(error)
} else {
    ...
}

Reply:
No I didn't. Thanks for informing me ;)